JS Tests

duggiemitchell opened the pull request:
todo: 
- [x] tests for filters/type.js
- [x] tests for filters/time.js
- [x] fix failing tests
- [x] possible refactor of filters

## Cute animal pic
![honeybadger](https://user-images.githubusercontent.com/12376362/74369836-1295bd80-4da4-11ea-8dab-b96136551ed1.jpg)


## Link to Trello card (if applicable)
https://trello.com/c/fer5NY7n


## Description
_The commit messages say what you did; this should explain why and how._
- [ ] Description is in Trello card



## Steps to test/reproduce
run `yarn test` to run tests in `./test` folder


## Reviewer tasks:

- [ ] Pulled branch and ran code locally (or viewed on preview deployment, if applicable)
- [ ] Reproduced the issue to review in the browser
- [ ] Reviewed code
- [ ] Updated Trello card (if applicable)

jgerigmeyer commented on package.json:
```diff
@@ -35,9 +35,11 @@
+    "test": "jest"
   },
   "devDependencies": {
+    "#": "link:./src/filters",
```
@duggiemitchell @mirisuzanne I added this link so we can import or require from e.g. `#/pages` instead of using the full path.

mirisuzanne replied:
Why `#` rather than  `~`?

jgerigmeyer replied:
Because it's specifically a shortcut to the 11ty filters, not to node modules -- I didn't want to confuse the two. 🤷‍♂ 

This is all arbitrary, of course. We could just use a generic shortcut to `src/` and do e.g. `require('#/filters/pages')` if we want.

duggiemitchell commented:
@jgerigmeyer or @wlonk 80% I could use some review on a couple problematic tests that are either failing or doesn't have coverage for.  I'm also not 100% sure on the best way to mock functions from outside libraries such as in the `/type.js` filters.

jgerigmeyer commented on src/filters/pages.js:
```diff
@@ -291,8 +291,7 @@ const pageType = (tags) => {
   ];
-
-  return tags ? tags.find((tag) => types.includes(tag)) : false;
+  return tags.filter((tag) => types.includes(tag));
```
This change is causing the build to fail, because 1) it is expecting a `string` return value instead of an array, and 2) `tags` might be missing entirely.

```suggestion
  return tags ? tags.find((tag) => types.includes(tag)) : false;
```

duggiemitchell replied:
OK.  Does it make sense for this to return an empty string instead of false if the tag isn't included in the list? (I think I may have made that change and the build is successful.)

jgerigmeyer replied:
Hm. I'm not sure practically, but in some ways `false` makes more sense to me theoretically. Or `undefined`?

jgerigmeyer commented on src/filters/tags.js:
```diff
@@ -43,9 +42,9 @@ params:
 const displayName = (tag) => {
   const capitalize = ([first, ...rest]) =>
-    first ? first.toUpperCase() + rest.join('') : '';
+    first && first.toUpperCase() + rest.join('');
```
Won't this change mean we now get `false` instead of `''` if tag is empty?

duggiemitchell replied:
Yup you are right.  Maybe I'm misunderstanding how this is expected to called; can't seem to get the coverage for returning an empty string :woman_shrugging:

jgerigmeyer commented on test/js/pages.test.js:
```diff
@@ -0,0 +1,165 @@
+  });
+
+  test('hasData', () => {
+    // this keeps returning undefined...
```
These are returning `undefined` because lodash is mocked out (I'm not sure why?). I think we should remove `jest.mock('lodash');` above.

jgerigmeyer commented on test/js/pages.test.js:
```diff
@@ -0,0 +1,165 @@
+    expect(pageKeys).toHaveProperty('sort');
+    // This is somehow failing but its in keys when printed //
+    expect(pageKeys).toHaveProperty('year');
+  });
```
You're already getting the list of keys with `Object.keys()`, so these should either be checking `expect(pageKeys).toContain('sort')` or `expect(testPage).toHaveProperty('sort')`.

jgerigmeyer commented on test/js/birds.test.js:
```diff
@@ -0,0 +1,87 @@
+
+describe('authorPage', () => {
+  test('returns author home page', () => {
+    authorPage(collection, 'erica'); // / this is returning undefined which is not expected
```
`authorPage` looks for a `data.bird` key on the page; the sample `collection` above does not have that.

jgerigmeyer commented on test/js/pages.test.js:
```diff
@@ -0,0 +1,165 @@
+  describe('pageType', () => {
+    test('Return one of several resource "types" from page tags', () => {
+      const tags = ['Workshops', 'Podcasts', 'foo', 'bar'];
+      expect(pageType(tags)).toEqual(['Workshops', 'Podcasts']);
```
This should just match the _first_ occurrence, not all matching ones.

duggiemitchell replied:
OK makes sense.  In `pageType` I feel like it would make more sense to return an empty string instead of false if the tag isn't included in the types list.

jgerigmeyer commented on test/js/tags.test.js:
```diff
@@ -0,0 +1,58 @@
+
+  test('tagData', () => {
+    // @todo needs a fix...
+    tagData(collection, tags, 'pageCount');
```
The first arg here is expected to be an object with all collections by tag.

jgerigmeyer commented:
> I'm also not 100% sure on the best way to mock functions from outside libraries such as in the `/type.js` filters.

@duggiemitchell Hm, that's a good question. Since those are relatively simple libraries, I'd probably start by _not_ mocking them at all. Then our tests are more similar to reality, even if we're also testing that those libraries do what we think they do.

duggiemitchell commented on src/filters/events.js:
```diff
@@ -76,8 +76,8 @@ params:
-const isFuture = (event) =>
-  event.end ? getDate(event.end) >= now : getDate(event.date) >= now;
+const isFuture = (page) =>
+  page.event ? getDate(page.event.end) >= now : getDate(page.date) >= now;
```
Yes, I believe so.  :+1:  can change it back to check for events...

duggiemitchell commented on src/filters/events.js:
```diff
@@ -98,6 +98,7 @@ params:
     type: array
     note: containing 11ty page objects
 */
+/* istanbul ignore next */
```
@mirisuzanne talked about changing how we filter events and ignoring this filter since it's not being used anywhere atm.

duggiemitchell commented on src/filters/pages.js:
```diff
@@ -193,13 +193,13 @@ params:
-  return page.data.renderData
-    ? page.data.renderData[key] || page.data[key]
-    : page.data[key];
+  return page.data.renderData ? page.data.renderData[key] : page.data[key];
```
@jgerigmeyer Makes sense, think I was reading this as either if `renderData` exists, return the page with matching key or the page.  Could have also been trying to get a test to pass and seems to work fine now after reverting the change.

jgerigmeyer commented on src/filters/pages.js:
```diff
@@ -193,13 +193,13 @@ params:
     type: string
 */
 const render = (page, key) => {
+  /* will pages ever not have data attributes? */
```
Ah, found it. https://www.11ty.dev/docs/collections/#collection-item-data-structure

So I guess pages _within a collection_ have stuff on `page.data`, but pages outside of a collection have their stuff on just `page`?

I have some concerns about the maintainability of the current structure, with a mix of Nunjucks and Markdown, lots and lots of JS filters and lots of Nunjucks macros. There are so many small pieces (which can be a plus) -- in practice it's very difficult for me to understand what any one piece is actually doing. It feels like lots of rabbit holes.

I'll try to think a bit more about what precisely is the issue I'm trying to articulate, and what I'd recommend to resolve it.

mirisuzanne replied:
Yeah, I find the difference between the two totally frustrating. I agree there's a lot going on. I'm not sure I know the fix…

duggiemitchell replied:
I will say it's not always clear to me what the data looks like (for things like pages and collections) and that sort of makes me feel the testing is brittle (if that makes sense). I may be worse at articulating what exactly my frustration is, so  I am interesting to hear you all's thoughts on how things are structured.

mirisuzanne replied:
There are two page objects, the link Jonny shred (a collection-page) and the current [`page` variable](https://www.11ty.dev/docs/data-eleventy-supplied/#page-variable-contents). They have a lot of overlap (`url`, `date`, `inputPath`, `outputPath`, `fileSlug`), but a few differences:

- collection-pages have `templateContent` and `data` (which includes page, directory, and global data)
- current `page` has `filePathStem`, but the content and data-properties are not included. We have to access them un-scoped as `content`, and `<property>`.

I find it very tempting to do `set page = collections.all | getPage(page.url)` first thing in our layouts, so that the page object includes all data and content. We would just have to make sure that layout is not used by any page excluded from collections. (or maybe there's no reason to exclude pages from collections…)

duggiemitchell commented on src/filters/pages.js:
```diff
@@ -254,12 +254,12 @@ params:
     note: optionally include events in sort/year dates
 */
 const byYear = (collection, events = true) => {
+  /* istanbul ignore else */
```
No reason I can think of now, removing...

duggiemitchell commented:
@jgerigmeyer I think I've addressed the review comments and have gone as far as I can without direction on a couple tests (especially related to `typopgraphy filters`) and I did make an external file to import collections, that may use some cleanup? (as in not representative of how collections actually look).

mirisuzanne commented on test/js/birds.test.js:
```diff
@@ -0,0 +1,35 @@
+});
+
+describe('active', () => {
+  test('filters pages from inactive birds', () => {
```
maybe clarify that these are not bird-authored pages, but bird-detail pages?

```suggestion
  test('filters inactive bird-detail pages', () => {
```

mirisuzanne commented on test/js/birds.test.js:
```diff
@@ -0,0 +1,35 @@
+
+    expect(active(collection2)).toEqual(expect.not.objectContaining(expected));
+  });
+  test('returns pages from inactive birds', () => {
```
```suggestion
  test('returns inactive bird-detail pages', () => {
```

jgerigmeyer commented:
There's more cleanup I'd like to do of course, but I'm good with this PR once @duggiemitchell and @mirisuzanne sign off 👍

jgerigmeyer commented:
![](https://github.trello.services/images/mini-trello-icon.png) [I want JS tests for back-end (eg filters) JS](https://trello.com/c/fer5NY7n/32-i-want-js-tests-for-back-end-eg-filters-js)
